[Modern Media Controls] add a way to show stats about the <video>

dcrousso opened the pull request:
#### e5887c759dce8d3396d463f6cc46ab0d7c177bf4
<pre>
[Modern Media Controls] add a way to show stats about the &lt;video&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=244101">https://bugs.webkit.org/show_bug.cgi?id=244101</a>
&lt;rdar://problem/98847210&gt;

Reviewed by Eric Carlson.

This will allow developers to gleam more information about the current state of `&lt;video&gt;` at a
glance, rather than having to use JS APIs to gather that info. This can also be expanded to shows
information that doesn&apos;t have an equivalent JS API.

* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
Add an experimental setting for this feature.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updatePreferences):
If this new experimental setting is enabled, also require `TrackConfigurationEnabled` as that&apos;s
needed for many of the stats to be shown.

* Source/WebCore/Modules/modern-media-controls/controls/media-controls.css:
(.stats-container): Added.
(.stats-container &gt; table): Added.
(.stats-container &gt; table &gt; tr &gt; :is(th, td)): Added.
(.stats-container &gt; table &gt; tr &gt; th): Added.
* Source/WebCore/Modules/modern-media-controls/media/media-controller.js:
(MediaController.prototype.setShowingStats): Added.
(MediaController.prototype.setShowingStats.createRow): Added.
When &quot;Show Media Stats&quot; is enabled, show a basic `&lt;table&gt;` containing rows for each stat that&apos;s
centered inside the `&lt;video&gt;`.

* Source/WebCore/Modules/modern-media-controls/media/overflow-support.js:
(OverflowSupport.prototype.syncControl):
* Source/WebCore/Modules/mediacontrols/MediaControlsHost.cpp:
(WebCore::MediaControlsHost::showMediaControlsContextMenu):
Also allow the (modern) media controls JS to provide `&quot;includeShowMediaStats&quot;`, which does the same
acts the same as the general context menu item.

* Source/WebCore/html/HTMLMediaElement.h:
(WebCore::HTMLMediaElement::showingStats const): Added.
The source of truth for &quot;are we showing media stats&quot; is in a member variable.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::setShowingStats): Added.
Invoke some JS to create/remove the `&lt;table&gt;`, updating the member variable depending on the result.

* Source/WebCore/platform/ContextMenuItem.h:
* Source/WebCore/platform/ContextMenuItem.cpp:
(WebCore::isValidContextMenuAction):
* Source/WebCore/page/ContextMenuController.h:
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::showContextMenu):
(WebCore::ContextMenuController::contextMenuItemSelected):
(WebCore::ContextMenuController::addDebuggingItems): Renamed from `addInspectElementItem`.
(WebCore::ContextMenuController::checkOrEnableIfNeeded const):
* Source/WebCore/rendering/HitTestResult.h:
* Source/WebCore/rendering/HitTestResult.cpp:
(WebCore::HitTestResult::toggleShowMediaStats const): Added.
(WebCore::HitTestResult::mediaStatsShowing const): Added.
* Source/WebKit/UIProcess/API/Cocoa/WKMenuItemIdentifiersPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKMenuItemIdentifiers.mm:
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::menuItemIdentifier):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(toTag):
Create a new `ContextMenuItemTagShowMediaStats`, adding it after `ContextMenuItemTagInspectElement`.

* Source/WebKit/UIProcess/ios/WKActionSheetAssistant.mm:
(-[WKActionSheetAssistant _uiMenuElementsForMediaControlContextMenuItems:]):
Add a way to create a separator in the (modern) media controls context menu on iOS.

* Source/WebCore/en.lproj/modern-media-controls-localized-strings.js:
* Source/WebCore/Modules/modern-media-controls/main.js:
(UIString):
Allow for multiple replacements in a single string.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::contextMenuItemTagShowMediaStats): Added.

Canonical link: <a href="https://commits.webkit.org/253583@main">https://commits.webkit.org/253583@main</a>
</pre>










<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/25103a9244fa558ba00f3aceb65f3fb7c9550b75

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86457 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/30377 "Built successfully") | [✅ ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/17424 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95302 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149013 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/90445 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/28778 "Built successfully") | [✅ ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25360 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/78610 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90546 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92073 "Passed tests") | [✅ ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/23353 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/73426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23404 "Passed tests") | 
| | [✅ ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/78361 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/17424 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66408 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/78393 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/26689 "Built successfully") | [✅ ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/17424 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/72029 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/26603 "Built successfully") | [✅ ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/17424 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/25745 "Passed tests") | 
| [✅ ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/2553 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28282 "Built successfully") | [✅ ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/73426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/74810 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28223 "Built successfully") | [✅ ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/17424 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16542 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -465,6 +465,14 @@ void HitTestResult::toggleMediaLoopPlayback() const
+void HitTestResult::toggleShowMediaStats() const
+{
+#if ENABLE(VIDEO)
+    if (HTMLMediaElement* mediaElt = mediaElement())
```
This isn't the first use of `mediaElt` in this file, but surely you can come up with a more descriptive variable name of the new code.

dcrousso replied:
lol yeah fair i can rename this

i didn't even really notice this cause it was mostly a copy pasta from "Loop" 😅

eric-carlson commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -518,6 +526,15 @@ bool HitTestResult::mediaLoopEnabled() const
+bool HitTestResult::mediaStatsShowing() const
+{
+#if ENABLE(VIDEO)
+    if (HTMLMediaElement* mediaElt = mediaElement())
```
Ditto

eric-carlson commented on Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
```diff
@@ -590,28 +587,23 @@ - (void)forwardContextMenuAction:(id)sender;
         ASSERT_NOT_REACHED();
+        break;
+
+    default:
+        break;
```
Why add a default??

dcrousso replied:
both cause i need to add a case for `ContextMenuItemTagShowMediaStats` and so that in the future if others add new items like that they don't even have to worry about handling it here

eric-carlson reviewed (approved):
r=me

dcrousso committed:
[Modern Media Controls] add a way to show stats about the <video>
https://bugs.webkit.org/show_bug.cgi?id=244101
<rdar://problem/98847210>

Reviewed by Eric Carlson.

This will allow developers to gleam more information about the current state of `<video>` at a
glance, rather than having to use JS APIs to gather that info. This can also be expanded to shows
information that doesn't have an equivalent JS API.

* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
Add an experimental setting for this feature.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updatePreferences):
If this new experimental setting is enabled, also require `TrackConfigurationEnabled` as that's
needed for many of the stats to be shown.

* Source/WebCore/Modules/modern-media-controls/controls/media-controls.css:
(.stats-container): Added.
(.stats-container > table): Added.
(.stats-container > table > tr > :is(th, td)): Added.
(.stats-container > table > tr > th): Added.
* Source/WebCore/Modules/modern-media-controls/media/media-controller.js:
(MediaController.prototype.setShowingStats): Added.
(MediaController.prototype.setShowingStats.createRow): Added.
When "Show Media Stats" is enabled, show a basic `<table>` containing rows for each stat that's
centered inside the `<video>`.

* Source/WebCore/Modules/modern-media-controls/media/overflow-support.js:
(OverflowSupport.prototype.syncControl):
* Source/WebCore/Modules/mediacontrols/MediaControlsHost.cpp:
(WebCore::MediaControlsHost::showMediaControlsContextMenu):
Also allow the (modern) media controls JS to provide `"includeShowMediaStats"`, which does the same
acts the same as the general context menu item.

* Source/WebCore/html/HTMLMediaElement.h:
(WebCore::HTMLMediaElement::showingStats const): Added.
The source of truth for "are we showing media stats" is in a member variable.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::setShowingStats): Added.
Invoke some JS to create/remove the `<table>`, updating the member variable depending on the result.

* Source/WebCore/platform/ContextMenuItem.h:
* Source/WebCore/platform/ContextMenuItem.cpp:
(WebCore::isValidContextMenuAction):
* Source/WebCore/page/ContextMenuController.h:
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::showContextMenu):
(WebCore::ContextMenuController::contextMenuItemSelected):
(WebCore::ContextMenuController::addDebuggingItems): Renamed from `addInspectElementItem`.
(WebCore::ContextMenuController::checkOrEnableIfNeeded const):
* Source/WebCore/rendering/HitTestResult.h:
* Source/WebCore/rendering/HitTestResult.cpp:
(WebCore::HitTestResult::toggleShowMediaStats const): Added.
(WebCore::HitTestResult::mediaStatsShowing const): Added.
* Source/WebKit/UIProcess/API/Cocoa/WKMenuItemIdentifiersPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKMenuItemIdentifiers.mm:
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::menuItemIdentifier):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(toTag):
Create a new `ContextMenuItemTagShowMediaStats`, adding it after `ContextMenuItemTagInspectElement`.

* Source/WebKit/UIProcess/ios/WKActionSheetAssistant.mm:
(-[WKActionSheetAssistant _uiMenuElementsForMediaControlContextMenuItems:]):
Add a way to create a separator in the (modern) media controls context menu on iOS.

* Source/WebCore/en.lproj/modern-media-controls-localized-strings.js:
* Source/WebCore/Modules/modern-media-controls/main.js:
(UIString):
Allow for multiple replacements in a single string.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::contextMenuItemTagShowMediaStats): Added.

Canonical link: https://commits.webkit.org/253583@main
